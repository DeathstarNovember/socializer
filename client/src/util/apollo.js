import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
