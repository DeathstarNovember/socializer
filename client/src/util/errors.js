import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
