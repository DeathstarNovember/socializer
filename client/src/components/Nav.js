import React, { useContext } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { Container, Nav, Navbar, NavDropdown } from "react-bootstrap";
import { Link } from "react-router-dom";
import renderIf from "render-if";
import { AuthContext } from "util/context";

export const GET_USER_INFO = gql`
  query GetUserInfo {
    currentUser {
      name
    }
  }
`;

const AuthNav = () => {
  const { loading, error, data } = useQuery(GET_USER_INFO);
  const { setAuth } = useContext(AuthContext);

  if (error) {
    if (
      error.graphQLErrors[0] &&
      error.graphQLErrors[0].message === "Unauthenticated"
    ) {
      setAuth(null);
    }
    return null;
  }

  if (data && !data.currentUser) {
    return null;
  }

  return (
    <NavDropdown
      title={loading ? "User" : data.currentUser.name}
      id="profile-dropdown"
    >
      <NavDropdown.Item onClick={() => setAuth(null)}>Log out</NavDropdown.Item>
    </NavDropdown>
  );
};

const AppNav = () => {
  const { token } = useContext(AuthContext);

  return (
    <Navbar bg="primary" variant="dark" className="mb-4" fixed="top">
      <Container>
        <Link to="/" className="navbar-brand">
          Socializer
        </Link>

        {renderIf(token)(
          <Nav>
            <Link to="/" className="nav-link" role="button">
              Feed
            </Link>
            <Link to="/chat" className="nav-link" role="button">
              Chat
            </Link>
          </Nav>,
        )}

        <Navbar.Toggle />
        <Navbar.Collapse className="justify-content-end">
          <Nav>
            {renderIf(token)(<AuthNav />)}

            {renderIf(!token)(
              <>
                <Link to="/login" className="nav-link" role="button">
                  Log in
                </Link>
                <Link to="/signup" className="nav-link" role="button">
                  Sign up
                </Link>
              </>,
            )}
          </Nav>
        </Navbar.Collapse>
      </Container>
    </Navbar>
  );
};

export default AppNav;
