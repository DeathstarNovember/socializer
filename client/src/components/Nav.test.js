import React from "react";
import { render, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "util/context";
import Nav, { GET_USER_INFO } from "./Nav";

describe("Nav", () => {
  it("renders correctly when unauthenticated", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ token: null, setAuth: jest.fn() }}>
          <MockedProvider mocks={[]} addTypename={false}>
            <Nav />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when loading", () => {
    const mocks = [
      {
        request: {
          query: GET_USER_INFO,
        },
        result: {
          data: {
            currentUser: {
              name: "John Doe",
            },
          },
        },
      },
    ];
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ token: "abc", setAuth: jest.fn() }}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <Nav />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when loaded", async () => {
    const mocks = [
      {
        request: {
          query: GET_USER_INFO,
        },
        result: {
          data: {
            currentUser: {
              name: "John Doe",
            },
          },
        },
      },
    ];
    const { container, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ token: "abc", setAuth: jest.fn() }}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <Nav />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    await wait(() => getByText("John Doe"));
    expect(container).toMatchSnapshot();
  });

  it("clears auth when loaded but token was stale", async () => {
    const mocks = [
      {
        request: {
          query: GET_USER_INFO,
        },
        result: {
          errors: [{ message: "Unauthenticated" }],
        },
      },
    ];
    const setAuth = jest.fn();
    render(
      <MemoryRouter>
        <AuthContext.Provider value={{ token: "abc", setAuth }}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <Nav />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    await wait(() => expect(setAuth).toBeCalledWith(null));
  });
});
