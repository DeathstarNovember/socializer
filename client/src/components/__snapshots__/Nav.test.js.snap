// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders correctly when loaded 1`] = `
<div>
  <nav
    class="mb-4 navbar navbar-expand navbar-dark bg-primary fixed-top"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Socializer
      </a>
      <div
        class="navbar-nav"
      >
        <a
          class="nav-link"
          href="/"
          role="button"
        >
          Feed
        </a>
        <a
          class="nav-link"
          href="/chat"
          role="button"
        >
          Chat
        </a>
      </div>
      <button
        aria-label="Toggle navigation"
        class="navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="justify-content-end navbar-collapse collapse"
      >
        <div
          class="navbar-nav"
        >
          <div
            class="dropdown nav-item"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle nav-link"
              href="#"
              role="button"
            >
              John Doe
            </a>
            <div
              aria-labelledby=""
              class="dropdown-menu"
            >
              <a
                class="dropdown-item"
                href="#"
                role="button"
              >
                Log out
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;

exports[`Nav renders correctly when loading 1`] = `
<div>
  <nav
    class="mb-4 navbar navbar-expand navbar-dark bg-primary fixed-top"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Socializer
      </a>
      <div
        class="navbar-nav"
      >
        <a
          class="nav-link"
          href="/"
          role="button"
        >
          Feed
        </a>
        <a
          class="nav-link"
          href="/chat"
          role="button"
        >
          Chat
        </a>
      </div>
      <button
        aria-label="Toggle navigation"
        class="navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="justify-content-end navbar-collapse collapse"
      >
        <div
          class="navbar-nav"
        />
      </div>
    </div>
  </nav>
</div>
`;

exports[`Nav renders correctly when unauthenticated 1`] = `
<div>
  <nav
    class="mb-4 navbar navbar-expand navbar-dark bg-primary fixed-top"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Socializer
      </a>
      <button
        aria-label="Toggle navigation"
        class="navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="justify-content-end navbar-collapse collapse"
      >
        <div
          class="navbar-nav"
        >
          <a
            class="nav-link"
            href="/login"
            role="button"
          >
            Log in
          </a>
          <a
            class="nav-link"
            href="/signup"
            role="button"
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
